biz.aQute.resolve/src/biz/aQute/resolve/ResolveProcess.java

```diff
@@ -369,17 +370,32 @@ private static void removeFrameworkAndInputResources(Map<Resource,List<Wire>> re
 				if (provider == requirer)
 					continue;
 
-				List<Wire> incoming = inverted.get(provider);
-				if (incoming == null) {
-					incoming = new LinkedList<Wire>();
-					inverted.put(provider, incoming);
+				List<Wire> incoming;
+				Resource capResource = wire.getCapability().getResource();
+				if (isFragment(capResource)) {
+					incoming = inverted.get(capResource);
+					if (incoming == null) {
+						incoming = new LinkedList<Wire>();
+						inverted.put(capResource, incoming);
+					}
+				}
+				else {
+					incoming = inverted.get(provider);
+					if (incoming == null) {
+						incoming = new LinkedList<Wire>();
+						inverted.put(provider, incoming);
+					}
```

Comment:
A simpler change may be:

```java
Resource capResource = wire.getCapability().getResource();
if (isFragment(capResource)) {
  provider = capResource;
}
```

before the `List<Wire> incoming =`

basically if the cap resource is a fragment, declare it the provider of the wire (which is what your longer code does via duplication).